##### **Understanding recursion** #####

##### Definition #####
- Function calling itself
- Almost all the situation where we use loops
  - substitute the loops using recursion
- Can solve problems that seem very complex at first

##### Example - factorial #####

In [1]:
def factorial(n):
    result = 1
    while n > 1:
        result = n * result
        n -= 1
    return result

In [2]:
factorial(5)

120

##### Example - identifying the base case #####
- Add a condition
  - ensures that our algorithm does't execute forever
- **Factorial base case** -> n = 1

In [4]:
# Avoid infinite recursion, the first thing is to identify the base case
def factorial_recursion(n):
    if n == 1:
        return 1
    else:
        return n * factorial_recursion(n-1) # recursive call

In [5]:
print(factorial_recursion(5))

120


In [7]:
%timeit factorial(5)

105 ns ± 0.204 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [8]:
%timeit factorial_recursion(5)

166 ns ± 0.518 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


##### How recursion works #####
- Computer uses a **stack** to keep track of the funtions
  - **Call stack**
- `factorial(5)` starts

- Before `factorial(5)` finishes ->\
  `factorial(4)` starts, \
  the computer needs to know that `factorial(5)` didn't finish, so it pushes the information into the call stack.

- Before `factorial(4)` finishes ->\
  `factorial(3)` starts

- Before `factorial(3)` finishes ->\
  `factorial(2)` starts

- Before `factorial(2)` finishes ->\
  `factorial(1)` starts

- `factorial(1)` finishes
  - returns 1

- `factorial(2)` finishes
  - returns 2

- `factorial(3)` finishes
  - returns 6

- `factorial(4)` finishes
  - returns 24

- `factorial(5)` finishes
  - returns 120
  


##### Dynamic programming #####
- Optimization technique
- Mainly applied to recursion
- Can reduce the complexity of recursive algorithms
- Used for:
  - Any problem that can be divided into smaller subproblems
  - Subproblems overlap
- Solutions of subproblems are saved, avoiding the need to recalculate
  - Memorization

In [9]:
def fibonacci(n):
  # Define the base case
  if n <= 1:
    return n
  else:
    # Call recursively to fibonacci
    return fibonacci(n-1) + fibonacci(n-2)
    
print(fibonacci(6))

8


In [10]:
# saving the solutions of the subproblems in the cache variable to avoid recalculating.

cache = [None]*(100)

def fibonacci(n): 
    if n <= 1:
        return n
    
    # Check if the value exists
    if not cache[n]:
        # Save the result in cache
        cache[n] = fibonacci(n-1) + fibonacci(n-2)
    
    return cache[n]
    

print(fibonacci(6))

8


In [11]:
print(cache[:10])

[None, None, 1, 2, 3, 5, 8, None, None, None]


##### Towers of Hanoi #####

In this exercise, you will implement the **Towers of Hanoi** puzzle with a **recursive algorithm**. The aim of this game is to transfer all the disks from one of the three rods to another, following these rules:

- You can only move one disk at a time.
- You can only take the upper disk from one of the stacks and place it on top of another stack.
- You cannot put a larger disk on top of a smaller one.

The algorithm shown is an implementation of this game with four disks and three rods called 'A', 'B' and 'C'. The code contains **two mistakes**. In fact, if you execute it, it crashes the console because it exceeds the maximum recursion depth. Can you find the bugs and fix them?

- Correct the base case.
- Correct the calls to the hanoi() function.

In [ ]:
# question

def hanoi(num_disks, from_rod, to_rod, aux_rod):
  # Correct the base case
  if num_disks >= 0:
    # Correct the calls to the hanoi function
    hanoi(num_disks, from_rod, aux_rod, to_rod)
    print("Moving disk", num_disks, "from rod", from_rod,"to rod",to_rod)
    hanoi(num_disks, aux_rod, to_rod, from_rod)   

num_disks = 4
source_rod = 'A'
auxiliar_rod = 'B'
target_rod = 'C'

hanoi(num_disks, source_rod, target_rod, auxiliar_rod)

##### Solution from DataCamp #####

In [1]:
def hanoi(num_disks, from_rod, to_rod, aux_rod):
  # Correct the base case
  if num_disks >= 1:
    # Correct the calls to the hanoi function
    hanoi(num_disks - 1, from_rod, aux_rod, to_rod)
    print("Moving disk", num_disks, "from rod", from_rod,"to rod",to_rod)
    hanoi(num_disks - 1, aux_rod, to_rod, from_rod)   

num_disks = 4
source_rod = 'A'
auxiliar_rod = 'B'
target_rod = 'C'

hanoi(num_disks, source_rod, target_rod, auxiliar_rod)

Moving disk 1 from rod A to rod B
Moving disk 2 from rod A to rod C
Moving disk 1 from rod B to rod C
Moving disk 3 from rod A to rod B
Moving disk 1 from rod C to rod A
Moving disk 2 from rod C to rod B
Moving disk 1 from rod A to rod B
Moving disk 4 from rod A to rod C
Moving disk 1 from rod B to rod C
Moving disk 2 from rod B to rod A
Moving disk 1 from rod C to rod A
Moving disk 3 from rod B to rod C
Moving disk 1 from rod A to rod B
Moving disk 2 from rod A to rod C
Moving disk 1 from rod B to rod C


##### Solution from Claude #####

In [2]:
def hanoi(num_disks, from_rod, to_rod, aux_rod):
    if num_disks == 1:
        print("Moving disk 1 from rod", from_rod, "to rod", to_rod)
        return
    
    hanoi(num_disks - 1, from_rod, aux_rod, to_rod)
    print("Moving disk", num_disks, "from rod", from_rod, "to rod", to_rod)
    hanoi(num_disks - 1, aux_rod, to_rod, from_rod)

num_disks = 4
source_rod = 'A'
auxiliar_rod = 'B'
target_rod = 'C'

hanoi(num_disks, source_rod, target_rod, auxiliar_rod)

Moving disk 1 from rod A to rod B
Moving disk 2 from rod A to rod C
Moving disk 1 from rod B to rod C
Moving disk 3 from rod A to rod B
Moving disk 1 from rod C to rod A
Moving disk 2 from rod C to rod B
Moving disk 1 from rod A to rod B
Moving disk 4 from rod A to rod C
Moving disk 1 from rod B to rod C
Moving disk 2 from rod B to rod A
Moving disk 1 from rod C to rod A
Moving disk 3 from rod B to rod C
Moving disk 1 from rod A to rod B
Moving disk 2 from rod A to rod C
Moving disk 1 from rod B to rod C
